tools/extra/qpafilter/qpafilter.py

```diff
@@ -0,0 +1,270 @@
+                        f'{DEGREES_C}).')
+            LOG.warning('Check QPA settings.')
+            return False
+
+        return True
+
+
+def get_verifier(category):
+    """Return the verifier class, given a category name."""
+    verifiers = { TEMPERATURE_CATEGORY: temp_verifier
+    }
+    return verifiers[category]
+
+
+class temp_filter:
+    """Calculate Warning Temperature thresholds, given the Fatal
+       value and supporting ratio."""
+    def __init__(self, args):
+        self.args = args
+
+    def filter(self, items):
+        for i in items:
+            fatal = float(i['fatal'])
+            units = i['units']
+
+            if 'override' in i:
+                override = float(i['override'])
+                warning = (override * fatal) / 100.0
+            else:
+                override = 0.0
+                warning = ((fatal * self.args.virt_warn_temp) /
+                          self.args.virt_fatal_temp)
+
+            i['warning'] = str(warning)
+
+            msg = (f'{i["label"]} warning: {i["warning"]} {units} '
+                   f'fatal: {i["fatal"]} {units}')
+            if override != 0.0:
+                msg += f' (override {override}%)'
+
+            # TODO: what to print here?
+            print(msg)
+
+        return True
+
+
+def get_filter(category):
+    """Return the filter class, given a category name."""
+    filters = { TEMPERATURE_CATEGORY: temp_filter
+    }
+    return filters[category]
+
+
+def read_qpa(in_file, temp_overrides):
+    """Read the input file and convert it to our data structure.
+       ie a dict keyed by the category name. The value for each
+       entry is a list of dictionaries with keys = {'label',
+       'fatal', 'warning', 'units'}.
+    """
+    override_d = {}
+    for ot in temp_overrides:
+        try:
+            olabel, opercentage = ot.split(':')
+        except ValueError:
+            LOG.warning(f'{ot} is not a valid temperature '
+                        f'override. Skipping')
+            continue
+        override_d[olabel] = opercentage
+
+    category_re = re.compile(CATEGORY_PATTERN, re.DOTALL|re.UNICODE)
+    item_re = re.compile(DATA_ITEM_PATTERN, re.UNICODE)
+    outer_d = defaultdict(list)
+    for mc in category_re.finditer(in_file.read()):
+        for mv in item_re.finditer(mc.group('values')):
+            label = mv.group('label').strip()
+            fatal = mv.group('value').strip()
+            units = mv.group('units').strip()
+
+            inner_d = {'label': label,
+                       'fatal': fatal,
+                       'units': units,
+                       'warning': 0.0 # placeholder
+                      }
+
+            if label in override_d:
+                inner_d['override'] = override_d[label]
+                LOG.info(f'Setting override for \'{label}\' to {inner_d["override"]}%')
+
+            outer_d[mc.group('category').strip()].append(inner_d)
```

Comment:
Can we move this to between lines 152,153? Similar to how you did lines 154-156.
```Python
category = mc.group('category').strip()
# ... ~line 167
    outer_d[category].append(inner_d)
```